### PR TITLE
6 bad ways to add type information to JSON

### PR DESCRIPTION
**Pull request by hilton**
New blog post. The paths are probably right for GitHub pages, but will need to change, so I’ll probably add a stub Jekyll set-up to make it possible to render this post as HTML.

Note that on the _Files changed_ view, you can select _View_ to view the [the blog post rendered as HTML](https://github.com/signavio/tech-blog/blob/deab1b4e7c15152c6c13c09a4b524b2b4096360a/_posts/2017-01-01-json-type-information.md).

### CODE REVIEW

File: _posts/2017-01-01-json-type-information.md [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+For example, instead of modeling an E.164 telephone number using a Java class, you could instead write code that checks the JSON string value when deserialising the JSON representation.
+
+Either way, either approach to using JSON that does not contain type information results in external assumptions about which types are present in the JSON structure.
+The next thing to try is making the type explicit.
+
+## 2. Inline type property
+
+For the next step you can specify each JSON object’s type in a special inline property.
+You can probably avoid name clashes by using an underscore prefix, calling it `_type`.
+In Java,
```
Comment (TimKam) [inline _posts/2017-01-01-json-type-information.md]: Is something missing here or should it be removed?

File: _posts/2017-01-01-json-type-information.md [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+```json
+"price" : {
+  "@Money" : {
+    "currency" : "EUR",
+    "amount" : 10.00
+  }
+}
+```
+
+However, the JSON still has an another level of nesting, albeit with less fluff.
```
Comment (TimKam) [inline _posts/2017-01-01-json-type-information.md]: *an another* -> *another* or *an additional* ?

File: _posts/2017-01-01-json-type-information.md [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+This approach also supports multiple inheritance, but you’ll have to decide for yourself whether you think that’s an advantage or disadvantage.
+
+## Schema definitions
+
+All of the above approaches refer to types by name.
+This means that you have to define the data types elsewhere, perhaps as a class model in an object-oriented programming language, using some API or in a declarative format.
+
+You might also consider options for JSON-centric data model definitions, that you can use to define types like `Money`:
+
+* [JSON schema](http://json-schema.org) - a JSON representation of type definitions that you use in JSON documents, with a similar focus to XML Schema.
```
Comment (TimKam) [inline _posts/2017-01-01-json-type-information.md]: *to* -> *as* ?

File: _posts/2017-01-01-json-type-information.md [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+---
+
+[JSON](https://en.wikipedia.org/wiki/JSON), specified by [RFC7159](https://tools.ietf.org/html/rfc7159) has taken over from [XML](https://en.wikipedia.org/wiki/XML) as data format of choice for web service [API](https://en.wikipedia.org/wiki/Application_programming_interface)s.
+Developers like how JSON is less verbose and easier to use than XML, but its limitations lead to extra work for some applications.
+
+![I don’t always use JSON, but when I do I include full type information](images/json-type-information-meme.jpg)
+
+JSON defines a small number of types: objects, arrays, numbers and strings, as well as three special values - `true`, `false` and `null`.
+However, JSON does not include a mechanism for defining custom domain types.
+Fortunately, there are several options to choose between, all bad.
```
Comment (TimKam) [inline _posts/2017-01-01-json-type-information.md]: This sentence sounds a bit strange to me, but I suppose it's on purpose.

File: _posts/2017-01-01-json-type-information.md [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+This blog post compares their disadvantages, so you can choose the least bad for your situation.
+
+## 1. No type information
+
+Consider this example of an invoice for some Signavio merchandise.
```
Comment (TimKam) [inline _posts/2017-01-01-json-type-information.md]: I guess writing  *Signavio* should facilitate brand awareness.
Perhaps we should write a couple of sentences as a guideline of how (not) to do this?

File: _posts/2017-01-01-json-type-information.md (hunk 3 above)
Comment (TimKam) [inline _posts/2017-01-01-json-type-information.md]: I think I just got the intended semantics of the sentence wrong at first glance.

**Comment by hilton** [comment]
@TimKam I’ve added a commit that links ‘Signavio’ to the web site. I don’t think anyone else needs to review - I figure we’re good as long as a blog post has one review.

**Pull request merged**